File: src/3d/scenedef.c

#include "scenedef.h"
#include "materials.h"

static struct Scene scene;
static struct Camera camera;
uint8_t n_lights = 0;

struct Scene* get_scene() {
    return &scene;
}

struct Camera* get_camera() {
    return &camera;
}

static void scene_add(struct Scene* s, struct Shape* obj) {
    s->objects.add(&s->objects, obj);
}

static void scene_add_light(struct Scene* s, struct Light light) {
    if(n_lights == MAX_LIGHTS) {
        return;
    }
    s->lights[n_lights] = light;
    n_lights++;
}

static void scene_set_light(struct Scene* s, uint8_t index, struct Light light) {
    if(index >= n_lights) {
        return;
    }
    s->lights[index] = light;
}

void init_scene() {
    init_materials();
    camera = make_camera();
    scene = (struct Scene){
        .camera = &camera,
        .objects = make_group(), 
        .background = make_pixel(1, 1, 1),
        .ambient = make_pixel(0.2, 0.2, 0.2),
        .reflections = 0,
        .shadows = false,
        .lights = {0},
        .add = scene_add,
        .add_light = scene_add_light,
        .set_light = scene_set_light,
    };
    scene.lights[0] = (struct Light){camera.eye, make_pixel(1.0, 1.0, 1.0)};
}

void free_scene() {
    for(uint16_t i=0; i < scene.objects.n_objects; i++) {
        struct Shape s = scene.objects.objects[i];
        switch(s.shape_type) {
            case SPHERE:
                free_sphere(&s.shape.s);
                break;
            case BOX:
                // don't need to free memory here
                break;
            default:
                // unreachable
                break;
        }
    }
}

File: src/3d/scenedef.zig
const std = @import("std");
const camera = @import("camera.zig");
const models = @import("models.zig");
const gfx_types = @import("gfx_types.zig");

pub const Scene = struct {
    const Self = @This();

    camera: *camera.Camera,
    objects: models.Group,
    background: gfx_types.Pixel,
    ambient: f32,
    shadows: bool,
    reflections: u8,
    lights: std.ArrayList(struct { gfx_types.Point(f32, 3), gfx_types.Pixel }),

    pub fn add(self: *Self, obj: *models.Shape) !void {
        try self.objects.add(obj.*);
    }
};

var scene: Scene = undefined;
var cam: camera.Camera = undefined;

pub fn init_scene(alloc: std.mem.Allocator) !void {
    cam = camera.Camera{};
    scene = Scene{
        .camera = &cam,
        .objects = try models.Group.init(alloc),
        .background = gfx_types.Pixel{ .r = 255, .g = 255, .b = 255 },
        .ambient = 0.2,
        .shadows = false,
        .reflections = 0,
        .lights = @FieldType(Scene, "lights").init(alloc),
    };
    try scene.lights.append(.{ cam.eye, .{ .r = 1, .g = 1, .b = 1 } });
}

pub fn deinit_scene(alloc: std.mem.Allocator) void {
    scene.objects.deinit(alloc);
}

pub fn get_scene() *Scene {
    return &scene;
}

pub fn get_camera() *camera.Camera {
    return &cam;
}
